From: "Paul E. McKenney" <paulmck@linux.vnet.ibm.com>
Date: Thu, 5 Sep 2013 17:02:11 -0700
Subject: rcu: Throttle invoke_rcu_core() invocations due to non-lazy callbacks
Git-commit: c337f8f58ed7cf150651d232af8222421a71463d
Patch-mainline: v3.13-rc1
References:

If a non-lazy callback arrives on a CPU that has previously gone idle
with no non-lazy callbacks, invoke_rcu_core() forces the RCU core to
run.  However, it does not update the conditions, which could result
in several closely spaced invocations of the RCU core, which in turn
could result in an excessively high context-switch rate and resulting
high overhead.

This commit therefore updates the ->all_lazy and ->nonlazy_posted_snap
fields to prevent closely spaced invocations.

Reported-by: Tibor Billes <tbilles@gmx.com>
Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
Tested-by: Tibor Billes <tbilles@gmx.com>
Reviewed-by: Josh Triplett <josh@joshtriplett.org>
Acked-by: Mike Galbraith <mgalbraith@suse.de>

---
 kernel/rcutree_plugin.h |    2 ++
 1 file changed, 2 insertions(+)

--- a/kernel/rcutree_plugin.h
+++ b/kernel/rcutree_plugin.h
@@ -1745,6 +1745,8 @@ static void rcu_prepare_for_idle(int cpu
 	 */
 	if (rdtp->all_lazy &&
 	    rdtp->nonlazy_posted != rdtp->nonlazy_posted_snap) {
+		rdtp->all_lazy = false;
+		rdtp->nonlazy_posted_snap = rdtp->nonlazy_posted;
 		invoke_rcu_core();
 		return;
 	}
